### ISSUE TITLE
[BUG]: FlutterSoundPlayer.onProgress position rollsback to 0 after a few seconds

### ISSUE THREAD

**Issue by feykro** [Canardoux/flutter_sound#1040]
## Flutter Sound Version : 

- **FULL** flavor, version 9.2.13 (held back by flutter neumorphic package)

- **Important**: Result of the command : ```flutter pub deps | grep flutter_sound```
- ├── flutter_sound 9.2.13
│   ├── flutter_sound_platform_interface 9.2.13
│   ├── flutter_sound_web 9.2.13
│   │   ├── flutter_sound_platform_interface...

----------------------------------------------------------

## Severity

- Does not cause crash but interferes with proper function

- Result is not what expected.

- App builds and runs correctly

- Important issue

--------------------------------------------------------

## Platforms you faced the error 

- Android (API 33)

- Emulator 

------------------------------------------------

**Describe the bug**
Subscription to a flutterSoundPlayer's `onProgress` Stream updates with a PlaybackDisposition which has a position that rollsback after the first few seconds of playback. For example, when logging the playback it will go 0.0 -> 0.1 -> 0.2 -> 0.0 and then behave normally.

**To Reproduce**
1. Create a player and open it
2. Set the subscriptionDuration : player.setSubscriptionDuration(const Duration(milliseconds: 100));
3. Set the onProgress stream and subscribe to it: playerSub = player.onProgress?.listen((e) {...});
4. Log "e" inside of that function
5. Start a file playback
6. Observe the logs to see the error
----------------------------------------------

# Logs!!!!
This is the code
```
playerSub = player.onProgress?.listen((e) {
      maxDuration = e.duration;
      position = e.position;
      sliderValue =(position.inMilliseconds / maxDuration.inMilliseconds) * 100;
      log("Max duration is: $maxDuration, current position is $position");
      log("Slider value: $sliderValue %");
      notifyListeners();
    });
```
And here is an example of the logs I get:
```
* here I start the playback*
log] Max duration is: 0:00:07.531000, current position is 0:00:00.000000
[log] Slider value: 0.0 %
[log] Max duration is: 0:00:07.531000, current position is 0:00:00.000000
[log] Slider value: 0.0 %
[log] Max duration is: 0:00:07.531000, current position is 0:00:00.012000
[log] Slider value: 0.15934138892577346 %
...
[log] Max duration is: 0:00:07.531000, current position is 0:00:02.541000
[log] Slider value: 33.74053910503253 %
[log] Max duration is: 0:00:07.531000, current position is 0:00:02.642000
[log] Slider value: 35.08166246182446 %
[log] Max duration is: 0:00:07.531000, current position is 0:00:00.183000
[log] Slider value: 2.4299561811180452 %
[log] Max duration is: 0:00:07.531000, current position is 0:00:00.284000
[log] Slider value: 3.7710795379099724 %
[log] Max duration is: 0:00:07.531000, current position is 0:00:00.385000
[log] Slider value: 5.112202894701899 %
...
[log] Max duration is: 0:00:07.531000, current position is 0:00:07.355000
[log] Slider value: 97.662992962422 %
[log] Max duration is: 0:00:07.531000, current position is 0:00:07.456000
[log] Slider value: 99.00411631921392 %
```
Slider value is basically a percentage indicating the progress of the playback. As you can see, the position value behaves normally for the first 2 seconds, and then its value rollsback to 0 and then goes back to progrssing normally.

-----------------------------------------------------------------

**Comment by Larpoux** [comment]
Thank you @feykro for this well documented issue.
I will try to have a look to it tomorrow, but I don’t guarantee anything because I don’t have much time to spend on flutter sound maintenance.

‘It would be fantastic if others people worked on this maintenance

**Comment by Larpoux** [comment]
@feykro : Do you use the ```pause()``` functions ?

**Comment by Larpoux** [comment]
I changed the way  the thread handles the timer function on Android : Flutter Sound 9.4.15

@feykro : Can you look if same, better or worst ?

**Comment by feykro** [comment]
Hello @Larpoux ,
First, I do use pause but this « rollback » happens in the normal case where I just do play and nothing else.
Second, I will try 9.4.15 but it seems like it’s incompatible with objectBox due to dependencies, which is a big issue.

**Comment by Larpoux** [comment]
Dependency on which module?

**Comment by Larpoux** [comment]
Now, Flutter Sound depends on ```flutter: '>=3.3.0'``` instead of ```flutter: ">=2.0.0"```.
Maybe its the problem.
I probably can downgrade to 3.0.0. I do not remember why I had to change this dependency. Probably because a dependency problem with the ```meta``` module.
Probably not a good idea to download to 2.0.0

**Comment by Larpoux** [comment]
In fact I have no reason to put a constraint on flutter version. I probably can remove the constraint.
Just tell me if this is your issue or if this is a dependency problem on another module.

**Comment by feykro** [comment]
Running on the latest stable version of flutter, with the latest version (4.0.1) of objectBox and 9.4.15, I get this:

```
Because objectbox_generator >=1.6.1-dev.0 <1.6.1 depends on objectbox 1.6.1-dev.0 and objectbox_generator >=1.6.2-dev.0 <1.6.2-dev.1 depends on objectbox 1.6.2-dev.0, objectbox_generator
      >=1.6.1-dev.0 <1.6.1 or >=1.6.2-dev.0 <1.6.2-dev.1 requires objectbox 1.6.1-dev.0 or 1.6.2-dev.0.
    And because objectbox_generator >=1.6.2-dev.1 <1.6.2-dev.2 depends on objectbox 1.6.2-dev.1, objectbox_generator >=1.6.1-dev.0 <1.6.1 or >=1.6.2-dev.0 <1.6.2-dev.2 requires objectbox
      1.6.1-dev.0 or 1.6.2-dev.0 or 1.6.2-dev.1.
    And because objectbox_generator >=1.6.2-dev.2 <1.6.2 depends on objectbox 1.6.2-dev.2 and objectbox_generator >=1.7.0-dev.0 <1.7.0-dev.1 depends on objectbox 1.7.0-dev.0,
      objectbox_generator >=1.6.1-dev.0 <1.6.1 or >=1.6.2-dev.0 <1.6.2 or >=1.7.0-dev.0 <1.7.0-dev.1 requires objectbox 1.6.1-dev.0 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.7.0-dev.0.
    And because objectbox_generator >=1.6.1 <1.6.2-dev.0 depends on objectbox 1.6.1 and objectbox_generator >=1.4.0-dev.1 <1.6.1-dev.0 depends on analyzer >=1.1.0 <4.0.0, objectbox_generator
      >=1.4.0-dev.1 <1.6.2-∞ or >=1.7.0-dev.0 <1.7.0-dev.1 requires objectbox 1.6.1-dev.0 or 1.6.1 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.7.0-dev.0 or analyzer >=1.1.0 <4.0.0.
    And because objectbox_generator >=1.2.0 <1.4.0-dev.1 depends on analyzer >=1.1.0 <3.0.0 and objectbox_generator >=0.13.0-nullsafety.0 <1.2.0 depends on analyzer ^1.1.0, objectbox_generator
      >=0.13.0-nullsafety.0 <1.6.2-∞ or >=1.7.0-dev.0 <1.7.0-dev.1 requires objectbox 1.6.1-dev.0 or 1.6.1 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.7.0-dev.0 or analyzer >=1.1.0 <4.0.0.
    And because build_runner >=2.4.6 depends on analyzer >=4.4.0 <7.0.0 and objectbox_generator >=1.7.0 <1.7.1-dev.0 depends on objectbox 1.7.0, if build_runner >=2.4.6 and objectbox_generator
      >=0.13.0-nullsafety.0 <1.6.2-∞ or >=1.7.0-dev.0 <1.7.0-dev.1 or >=1.7.0 <1.7.1-dev.0 then objectbox 1.6.1-dev.0 or 1.6.1 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.7.0-dev.0 or 1.7.0.
    And because objectbox_generator >=1.6.2 <1.7.0-dev.0 depends on objectbox 1.6.2 and objectbox_generator >=1.7.0-dev.1 <1.7.0 depends on objectbox 1.7.0-dev.1, if build_runner >=2.4.6 and
      objectbox_generator >=0.13.0-nullsafety.0 <1.7.1-dev.0 then objectbox 1.6.1-dev.0 or 1.6.1 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.6.2 or 1.7.0-dev.0 or 1.7.0-dev.1 or 1.7.0.
(1) So, because objectbox_generator <0.13.0-nullsafety.0 doesn't support null safety and objectbox_generator >=1.7.1-dev.0 depends on cryptography ^2.0.5, if build_runner >=2.4.6 and
      objectbox_generator any then objectbox 1.6.1-dev.0 or 1.6.1 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.6.2 or 1.7.0-dev.0 or 1.7.0-dev.1 or 1.7.0 or cryptography ^2.0.5.

    Because cryptography >=2.0.0 <2.1.0 depends on js ^0.6.3 and cryptography >=2.1.0 <2.6.0 depends on js ^0.6.2, cryptography >=2.0.0 <2.6.0 requires js ^0.6.2.
    And because cryptography >=2.6.0 depends on js ^0.6.7, cryptography >=2.0.0 requires js ^0.6.2.
    And because if build_runner >=2.4.6 and objectbox_generator any then objectbox 1.6.1-dev.0 or 1.6.1 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.6.2 or 1.7.0-dev.0 or 1.7.0-dev.1 or
      1.7.0 or cryptography ^2.0.5 (1), if build_runner >=2.4.6 and objectbox_generator any then objectbox 1.6.1-dev.0 or 1.6.1 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.6.2 or 1.7.0-dev.0
      or 1.7.0-dev.1 or 1.7.0 or js ^0.6.2.
    And because flutter_sound >=9.4.15 depends on flutter_sound_web 9.4.15 which depends on js ^0.7.1, if build_runner >=2.4.6 and objectbox_generator any and flutter_sound >=9.4.15 then
      objectbox 1.6.1-dev.0 or 1.6.1 or 1.6.2-dev.0 or 1.6.2-dev.1 or 1.6.2-dev.2 or 1.6.2 or 1.7.0-dev.0 or 1.7.0-dev.1 or 1.7.0.
    And because reatgistre depends on both flutter_sound ^9.4.15 and objectbox ^4.0.1, build_runner >=2.4.6 is incompatible with objectbox_generator.
    So, because reatgistre depends on both build_runner ^2.4.9 and objectbox_generator any, version solving failed.

The lower bound of "sdk: '>=2.5.0 <3.0.0'" must be 2.12.0 or higher to enable null safety.
For details, see https://dart.dev/null-safety


You can try one of the following suggestions to make the pubspec resolve:
* Consider downgrading your constraint on objectbox: flutter pub add objectbox:^1.7.0
* Consider downgrading your constraint on flutter_sound: flutter pub add flutter_sound:^9.2.13
```

**Comment by Larpoux** [comment]
I released Flutter Sound 9.4.16 : 

### 9.4.16
Downgrade the js dependency in flutter_sound_web pubspec.yaml to 
```js: ^0.6.2``` instead of ```js: ^0.7.1```. 
Not sure this is good. 

I did that to be nice for you, but in fact I think that it is a problem with objectbox and not with Flutter Sound. Perhaps you could contact objectbox maintener.

**Comment by feykro** [comment]
Hello,
Thanks a ton, I was about to fork the repo to try and do this myself.
I've investigated the issue, I forked ObjectBox and upgrading the dependency `cryptography` to latest in order to get on `js 0.7.1` caused even more compatibility issues. I'll get in touch with the guys there and comeback to you on that matter as soon as I have news.

As for the FlutterSoundPlayer, I think this little dependency change broke something. When I run
```
player.startPlayer(
            codec: Codec.defaultCodec,
            fromURI: recording.filePath,
            whenFinished: () {
              playerState = ModalPlayerState.stopped;
              notifyListeners();
            });
```
(note that this function is not asynchronous)
I get a crash with:
```
┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): │ #0   FlutterSoundPlayer._startPlayer (package:flutter_sound/public/flutter_sound_player.dart:775:13)
I/flutter ( 4926): │ #1   FlutterSoundPlayer.startPlayer.<anonymous closure> (package:flutter_sound/public/flutter_sound_player.dart:755:17)
I/flutter ( 4926): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter ( 4926): │ 🐛 FS:---> startPlayer 
I/flutter ( 4926): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): │ #0   FlutterSoundPlayer._stop (package:flutter_sound/public/flutter_sound_player.dart:1108:13)
I/flutter ( 4926): │ #1   FlutterSoundPlayer._startPlayer (package:flutter_sound/public/flutter_sound_player.dart:802:11)
I/flutter ( 4926): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter ( 4926): │ 🐛 FS:---> _stop 
I/flutter ( 4926): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): │ #0   FlutterSoundPlayer._stop (package:flutter_sound/public/flutter_sound_player.dart:1135:13)
I/flutter ( 4926): │ #1   <asynchronous suspension>
I/flutter ( 4926): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter ( 4926): │ 🐛 FS:<--- _stop 
I/flutter ( 4926): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): │ #0   FlutterSoundPlayer.stopPlayerCompleted (package:flutter_sound/public/flutter_sound_player.dart:294:13)
I/flutter ( 4926): │ #1   MethodChannelFlutterSoundPlayer.channelMethodCallHandler.<anonymous closure> (package:flutter_sound_platform_interface/method_channel_flutter_sound_player.dart:103:21)
I/flutter ( 4926): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter ( 4926): │ 🐛 ---> stopPlayerCompleted: true
I/flutter ( 4926): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): │ #0   FlutterSoundPlayer.stopPlayerCompleted (package:flutter_sound/public/flutter_sound_player.dart:309:13)
I/flutter ( 4926): │ #1   MethodChannelFlutterSoundPlayer.channelMethodCallHandler.<anonymous closure> (package:flutter_sound_platform_interface/method_channel_flutter_sound_player.dart:103:21)
I/flutter ( 4926): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter ( 4926): │ 🐛 <--- stopPlayerCompleted: true
I/flutter ( 4926): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): ┌───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
I/flutter ( 4926): │ #0   FlutterSoundPlayer._startPlayer (package:flutter_sound/public/flutter_sound_player.dart:827:13)
I/flutter ( 4926): │ #1   <asynchronous suspension>
I/flutter ( 4926): ├┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄┄
I/flutter ( 4926): │ 🐛 FS:<--- startPlayer 
I/flutter ( 4926): └───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
E/AndroidRuntime( 4926): FATAL EXCEPTION: Timer-0
E/AndroidRuntime( 4926): Process: com.example.reatgistre, PID: 4926
E/AndroidRuntime( 4926): java.lang.RuntimeException: Methods marked with @UiThread must be executed on the main thread. Current thread: Timer-0
E/AndroidRuntime( 4926): 	at io.flutter.embedding.engine.FlutterJNI.ensureRunningOnMainThread(FlutterJNI.java:1493)
E/AndroidRuntime( 4926): 	at io.flutter.embedding.engine.FlutterJNI.dispatchPlatformMessage(FlutterJNI.java:1109)
E/AndroidRuntime( 4926): 	at io.flutter.embedding.engine.dart.DartMessenger.send(DartMessenger.java:281)
E/AndroidRuntime( 4926): 	at io.flutter.embedding.engine.dart.DartExecutor$DefaultBinaryMessenger.send(DartExecutor.java:464)
E/AndroidRuntime( 4926): 	at io.flutter.embedding.engine.dart.DartExecutor.send(DartExecutor.java:217)
E/AndroidRuntime( 4926): 	at io.flutter.plugin.common.MethodChannel.invokeMethod(MethodChannel.java:117)
E/AndroidRuntime( 4926): 	at io.flutter.plugin.common.MethodChannel.invokeMethod(MethodChannel.java:102)
E/AndroidRuntime( 4926): 	at xyz.canardoux.fluttersound.FlutterSoundManager.invokeMethod(FlutterSoundManager.java:49)
E/AndroidRuntime( 4926): 	at xyz.canardoux.fluttersound.FlutterSoundSession.log(FlutterSoundSession.java:128)
E/AndroidRuntime( 4926): 	at xyz.canardoux.TauEngine.FlautoPlayer.logDebug(FlautoPlayer.java:577)
E/AndroidRuntime( 4926): 	at xyz.canardoux.TauEngine.FlautoPlayer$3.run(FlautoPlayer.java:350)
E/AndroidRuntime( 4926): 	at java.util.TimerThread.mainLoop(Timer.java:563)
E/AndroidRuntime( 4926): 	at java.util.TimerThread.run(Timer.java:513)
I/Process ( 4926): Sending signal. PID: 4926 SIG: 9
```

Since it seems like you're not maintaining the package anymore, I would totally understand if you just wanted to revert the change and leave it be.

**Comment by Larpoux** [comment]
OK.
My change about the thread calling the ```onProgress``` callback was a bad idea. This is the reason of your new crash.
I will revert this change and look a little further to see if I understand your progression issue during play().

For the dependency problem, I suggest to retain the downgrade that I did on the JS dependency. It would be probably better for you to use the official plugin. If nobody complains about this JS dependency, this is OK for me.

**Comment by Larpoux** [comment]
I released Flutter Sound 9.4.167: 

### 9.4.17
The patch ### 9.4.15 was very bad. We have crash during the callback. Reverts the 9.4.15 patch

**Comment by feykro** [comment]
Final update, the issue seems to be related to Android's native call: https://github.com/bluefireteam/audioplayers/issues/1324

**Comment by Larpoux** [comment]
@feykro :  I released Flutter Sound 9.4.18.
There is now a patch inside the Android Timer so that the position never decreases.
This is not perfect but I cannot do better because there is clearly a bug inside Google Android MediaPlayer.
